cmd/reconciler/start/service/component_integration_suite_test.go

```diff
@@ -0,0 +1,313 @@
+		callbackMockPort: 11111,
+
+		workerConfig: &cliRecon.WorkerConfig{
+			Timeout: 35 * time.Second,
+			Workers: 2,
+		},
+
+		reconcilerPort: 9999,
+
+		kubeClientConfig: &k8s.Config{
+			MaxRetries:       3,
+			ProgressInterval: 5 * time.Second,
+			ProgressTimeout:  16 * time.Second,
+			RetryDelay:       2 * time.Second,
+		},
+	})
+}
+
+func (s *reconcilerIntegrationTestSuite) SetupSuite() {
+	s.containerSuite.SetupSuite()
+	s.serverStartMutex = sync.Mutex{}
+	s.testLogger = logger.NewLogger(true)
+
+	//use ./test folder as workspace
+	wsf, err := chart.NewFactory(nil, s.testDirectory, s.testLogger)
+	s.NoError(err)
+
+	// TODO this currently blocks parallel execution as tests share one execution directory
+	s.NoError(service.UseGlobalWorkspaceFactory(wsf))
+	s.chartFactory = wsf
+
+	kubeConfig := test.ReadKubeconfig(s.T())
+
+	//create kubeClient (e.g. needed to verify reconciliation results)
+	kubeClient, k8sErr := k8s.NewKubernetesClient(kubeConfig, s.testLogger, s.kubeClientConfig)
+	s.NoError(k8sErr)
+
+	s.kubeClient = kubeClient
+
+	cliOptions := &cli.Options{Verbose: true}
+	cliOptions.Registry, err = persistency.NewRegistry(s.containerSuite, true)
+	s.NoError(err)
+	o := cliRecon.NewOptions(cliOptions)
+	o.ServerConfig.Port = s.reconcilerPort
+	o.WorkerConfig = s.workerConfig
+	o.HeartbeatSenderConfig.Interval = 2 * time.Second
+	o.ProgressTrackerConfig.Interval = 1 * time.Second
+	o.RetryConfig.RetryDelay = 2 * time.Second
+	o.RetryConfig.MaxRetries = 3
+	s.options = o
+}
+
+func (s *reconcilerIntegrationTestSuite) TearDownSuite() {
+	s.containerSuite.TearDownSuite()
+}
+
+func (s *reconcilerIntegrationTestSuite) startAndWaitForComponentReconciler(settings *componentReconcilerIntegrationSettings) {
+	recon, reconErr := service.NewComponentReconciler(settings.name)
+	s.NoError(reconErr)
+	recon = recon.Debug()
+
+	if settings.reconcileAction != nil {
+		recon = recon.WithReconcileAction(settings.reconcileAction)
+	}
+
+	s.T().Cleanup(func() {
+		// this cleanup runs with Helm so it can happen that the cleanup unblocks while the finalizers are still not
+		// dropped for a namespace. This can cause test failures for successive tests using the same namespace as the
+		// API server will reject the request. You can circumvent this by using a different namespace.
+		cleanUp := service.NewTestCleanup(recon, s.kubeClient)
+
+		// this is needed to make sure the helm chart can be found under the right version
+		version := settings.version
+		if settings.url != "" {
+			version = chart.GetExternalArchiveComponentHashedVersion(settings.url, settings.name)
+		}
+
+		cleanUp.RemoveKymaComponent(
+			s.T(),
+			version,
+			settings.name,
+			settings.namespace,
+			settings.url,
+		)
+
+		// this is done as Cleanup of the externally downloaded component
+		if settings.deleteAfterTest {
+			s.NoError(s.chartFactory.Delete(version))
+		}
+	})
+
+	componentReconcilerServerContext, cancel := context.WithCancel(s.testContext)
```

Comment:
I think use `context.WithTimeout(parent Context, timeout time.Duration)` will do the same purpose and avoid write time.Sleep in the `s.T().Cleanup()`

Reply:
This does not work as WithTimeout will start the timer when creating the context. What we want instead is starting the timeout after cancel was called. I tried this and you can see that the server immediately shuts down at https://status.build.kyma-project.io/view/gs/kyma-prow-logs/pr-logs/pull/kyma-incubator_reconciler/956/pre-main-reconciler-integration-k3d/1510998581635452928#1:build-log.txt%3A643-654

Reply:
aha, sorry, then I understood in wrong direction..